core/services/ocr2/plugins/ccip/tokendata/bgworker.go

```diff
@@ -0,0 +1,237 @@
+package tokendata
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal"
+)
+
+type msgResult struct {
+	TokenAmountIndex int
+	Err              error
+	Data             []byte
+}
+
+type Worker interface {
+	// AddJobsFromMsgs will include the provided msgs for background processing.
+	AddJobsFromMsgs(ctx context.Context, msgs []internal.EVM2EVMOnRampCCIPSendRequestedWithMeta)
+
+	// GetMsgTokenData returns the token data for the provided msg. If data are not ready it keeps waiting
+	// until they get ready. Important: Make sure to pass a proper context with timeout.
+	GetMsgTokenData(ctx context.Context, msg internal.EVM2EVMOnRampCCIPSendRequestedWithMeta) ([][]byte, error)
+
+	GetReaders() map[common.Address]Reader
+}
+
+type BackgroundWorker struct {
+	tokenDataReaders map[common.Address]Reader
+	numWorkers       int
+	jobsChan         chan internal.EVM2EVMOnRampCCIPSendRequestedWithMeta
+	resultsCache     *resultsCache
+	timeoutDur       time.Duration
+}
+
+func (w *BackgroundWorker) AddJobsFromMsgs(ctx context.Context, msgs []internal.EVM2EVMOnRampCCIPSendRequestedWithMeta) {
+	go func() {
+		for _, msg := range msgs {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				if len(msg.TokenAmounts) > 0 {
+					w.jobsChan <- msg
+				}
+			}
+		}
+	}()
+}
+
+func (w *BackgroundWorker) GetReaders() map[common.Address]Reader {
+	return w.tokenDataReaders
+}
+
+func (w *BackgroundWorker) GetMsgTokenData(ctx context.Context, msg internal.EVM2EVMOnRampCCIPSendRequestedWithMeta) ([][]byte, error) {
+	res, err := w.getMsgTokenData(ctx, msg.SequenceNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenDatas := make([][]byte, len(msg.TokenAmounts))
+	for _, r := range res {
+		if r.Err != nil {
+			return nil, r.Err
+		}
+		if r.TokenAmountIndex < 0 || r.TokenAmountIndex >= len(msg.TokenAmounts) {
+			return nil, fmt.Errorf("token data index incosistency")
```

Comment:
Typo: `inconsistency`

```suggestion
			return nil, fmt.Errorf("token data index inconsistency")
```